linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -3137,20 +3137,69 @@ private class FunctionEmitter private (
   private def genArrayValue(tree: ArrayValue): Type = {
     val ArrayValue(arrayTypeRef, elems) = tree
 
-    val expectedElemType = arrayTypeRef match {
-      case ArrayTypeRef(base: PrimRef, 1) => base.tpe
-      case _                              => AnyType
-    }
-
-    // Mark the position for the header of `genArrayValue`
     markPosition(tree)
 
-    SWasmGen.genArrayValue(fb, arrayTypeRef, elems.size) {
-      // Create the underlying array
-      elems.foreach(genTree(_, expectedElemType))
+    arrayTypeRef match {
+      case ArrayTypeRef(base: PrimRef, 1) if elems.forall(_.isInstanceOf[Literal]) =>
+        // Use a constant array in a data segment
+        val length = elems.size
 
-      // Re-mark the position for the footer of `genArrayValue`
-      markPosition(tree)
+        val (dataID, offset) = (base.charCode: @switch) match {
+          case 'Z' =>
+            ctx.constantArrayPool.addArray8(elems) { (buffer, elem) =>
```

Comment:
I feel this interface is a bit roundabout, but I couldn't find anything better :shrug:

Reply:
I agree, it's not a great interface, but likewise this is the least bad I came up with.